hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsOperatorUtil.java

```diff
@@ -0,0 +1,123 @@
+    }
+
+    public static MetricsOperator cast(Object operatorCandidate) {
+        if (operatorCandidate instanceof MetricsOperator) {
+            return (MetricsOperator) operatorCandidate;
+        } else {
+            return DUMMY_OP;
+        }
+    }
+
+    public static <T, R> FunctionEx<T, R> wrap(FunctionEx<T, R> functionEx, Object operatorCandidate) {
+        if (operatorCandidate instanceof MetricsOperator) {
+            return new WrappedFunctionEx<>(functionEx, Collections.singletonList((MetricsOperator) operatorCandidate));
+        } else {
+            return functionEx;
+        }
+    }
+
+    public static <T, R> FunctionEx<T, R> wrapAll(FunctionEx<T, R> functionEx, List<?> operatorCandidates) {
+        List<MetricsOperator> operators = operatorCandidates.stream()
+                .filter(MetricsOperator.class::isInstance)
+                .map(MetricsOperator.class::cast)
+                .collect(toList());
+        if (operators.isEmpty()) {
+            return functionEx;
+        } else {
+            return new WrappedFunctionEx<>(functionEx, operators);
+        }
+    }
+
+    public static <P> SupplierEx<P> wrap(SupplierEx<P> supplierEx, Object operatorCandidate) {
+        if (operatorCandidate instanceof MetricsOperator) {
+            return new WrappedSupplierEx<>(supplierEx, Collections.singletonList((MetricsOperator) operatorCandidate));
+        } else {
+            return supplierEx;
+        }
+    }
+
+    private static class WrappedFunctionEx<T, R> implements FunctionEx<T, R>, MetricsOperator, Serializable {
+
+        private final FunctionEx<T, R> functionEx;
+        private final List<MetricsOperator> metricsOperators;
+
+        WrappedFunctionEx(FunctionEx<T, R> functionEx, List<MetricsOperator> metricsOperators) {
+            this.functionEx = functionEx;
+            this.metricsOperators = metricsOperators;
+        }
+
+        @Override
+        public R applyEx(T t) throws Exception {
+            return functionEx.applyEx(t);
+        }
+
+        @Override
+        public void init(MetricsContext context) {
+            for (MetricsOperator operator : metricsOperators) {
+                operator.init(context);
+            }
+        }
+    }
+
+    private static class WrappedSupplierEx<P> implements SupplierEx<P>, MetricsOperator, Serializable {
+
+        private final SupplierEx<P> supplierEx;
+        private final List<MetricsOperator> metricsOperators;
+
+        WrappedSupplierEx(SupplierEx<P> supplierEx, List<MetricsOperator> metricsOperators) {
+            this.supplierEx = supplierEx;
+            this.metricsOperators = metricsOperators;
+        }
+
+        @Override
+        public P getEx() throws Exception {
+            return supplierEx.getEx();
+        }
+
+        @Override
+        public void init(MetricsContext context) {
+            for (MetricsOperator operator : metricsOperators) {
+                operator.init(context);
```

Comment:
wouldn't they overwrite each other? also wouldn't it be better to merge the counters in one, and still keep one of the "operator" (to be renamed) object.

Reply:
Sorry, but this doesn't make sense to me, our mental models diverge somewhere. We should talk.

Reply:
I mean if two lambdas provide metrics with same name, they would be overwriten?

Reply:
Metrics with the same name would be definitely overwritten, because eventually they will end up in `MetricsRegistryImplementation.registerInternal`. The overwrite would be logged but that's it. That being said, once I figure out how to wrap counters and use lambdas I'll take another look if I could improve something on my side (ie. above code being commented on).